app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,61 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
+
+  def new; end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = t ".info_flash"
+      redirect_to root_url
+    else
+      flash.now[:danger] = t ".danger_flash"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?                  # Case (3)
+      @user.errors.add :password, t(".error_password")
+      render :edit
+    elsif @user.update_attributes(user_params)          # Case (4)
+      log_in @user
+      @user.update_attribute :reset_digest, nil
+      flash[:success] = t ".success_flash"
+      redirect_to @user
+    else
+      render :edit # Case (2)
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
+  end
+
+  # Confirms a valid user.
+  def valid_user
+    unless @user&.activated? &&
```

Comment:
`&.` đi nhé, vì đảm bảo tới method này user đã khác nil rồi

Reply:
> `&.` đi nhé, vì đảm bảo tới method này user đã khác nil rồi

e có bổ sung thông báo lỗi ở hàm này. e k biết phải thông báo lỗi như thế nào. c giúp e với ạ.

Reply:
to box chung, nói rõ hơn chú mún thông báo cái chi, hoặc lên bàn chị

Reply:
> to box chung, nói rõ hơn chú mún thông báo cái chi, hoặc lên bàn chị

e muốn thông báo lỗi ở hàm valid_user.
trường hợp sử dụng lại link reset pass cũ hoặc digest sai

Reply:
đã xong ạ